src/linear_algebra/char_poly/coeff.lean

```diff
@@ -0,0 +1,191 @@
+  rw ← sub_add_cancel (∏ (i : n), (X - C (M i i))) (char_poly M) at mon,
+  rw monic at *, rw leading_coeff_add_of_degree_lt at mon, rw ← mon,
+  rw degree_char_poly_eq_dim, rw ← neg_sub, rw degree_neg,
+  apply lt_trans (low_degree_char_poly_sub_diagonal M), rw with_bot.coe_lt_coe,
+  rw ← nat.pred_eq_sub_one, apply nat.pred_lt, apply h,
+end
+
+lemma char_poly_monic (M : matrix n n R) :
+  monic (char_poly M) :=
+begin
+  classical, by_cases h : nontrivial R,
+  { letI := h, apply char_poly_monic_of_nontrivial, },
+  { rw nontrivial_iff at h, push_neg at h, apply h, }
+end
+
+--shouldn't need these instances, but might need casework
+theorem trace_from_char_poly [nontrivial R] [nonempty n] (M: matrix n n R) :
+(matrix.trace n R R) M = -(char_poly M).coeff (fintype.card n - 1) :=
+begin
+  rw high_coeff_char_poly_eq_coeff_prod_diag, swap, refl,
+  rw [fintype.card, prod_X_sub_C_coeff_card_pred univ (λ i : n, M i i)], simp,
+  rw [← fintype.card, fintype.card_pos_iff], apply_instance,
+end
+
+lemma ring_hom_det {S : Type w} [comm_ring S] {M : matrix n n R} {f : R →+* S} :
+  f M.det = matrix.det (f.map_matrix M) :=
+by simp [matrix.det, f.map_sum, f.map_prod]
+
+lemma alg_hom_det {S : Type w} [comm_ring S] [algebra R S] {T : Type z} [comm_ring T] [algebra R T]
+  {M : matrix n n S} {f : S →ₐ[R] T} :
+  f M.det = matrix.det ((f : S →+* T).map_matrix M) :=
+by rw [← alg_hom.coe_to_ring_hom, ring_hom_det]
+
+lemma matrix.scalar.commute (r : R) (M : matrix n n R) : commute (scalar n r) M :=
+by simp [commute, semiconj_by]
+
+-- I feel like this should use polynomial.alg_hom_eval₂_algebra_map
+lemma eval_mat_poly_equiv (M : matrix n n (polynomial R)) (r : R) (i j : n) :
+  polynomial.eval r (M i j) = polynomial.eval ((scalar n) r) (mat_poly_equiv M) i j :=
+begin
+  unfold polynomial.eval, unfold eval₂,
+  transitivity finsupp.sum (mat_poly_equiv M) (λ (e : ℕ) (a : matrix n n R),
+    (a * (scalar n) r ^ e) i j),
+  { simp_rw ← (scalar n).map_pow, simp_rw ← (matrix.scalar.commute _ _).eq,
+    simp only [coe_scalar, matrix.one_mul, ring_hom.id_apply,
+      smul_val, mul_eq_mul, algebra.smul_mul_assoc],
+    have h : ∀ x : ℕ, (λ (e : ℕ) (a : R), r ^ e * a) x 0 = 0 := by simp,
+    rw ← finsupp.sum_map_range_index h, swap, refl,
+    refine congr (congr rfl _) (by {ext, rw mul_comm}), ext, rw finsupp.map_range_apply,
+    simp [apply_eq_coeff], },
+  { unfold finsupp.sum, rw sum_apply, rw sum_apply, dsimp, refl, }
+end
+
+lemma eval_det (M : matrix n n (polynomial R)) (r : R) :
+  polynomial.eval r M.det = (polynomial.eval (matrix.scalar n r) (mat_poly_equiv M)).det :=
+begin
+  rw [polynomial.eval, ← coe_eval₂_ring_hom, ring_hom_det],
+  apply congr_arg det, ext, convert eval_mat_poly_equiv _ _ _ _,
+end
+
+theorem det_from_char_poly (M: matrix n n R) :
+M.det = (-1)^(fintype.card n) * (char_poly M).coeff 0:=
+by rw [coeff_zero_eq_eval_zero, char_poly, eval_det, mat_poly_equiv_char_matrix, ← det_smul]; simp
```

Comment:
Use `by { X, Y }` in favour of `by X; Y`, unless there are actually multiple goals being passed to `Y`.